### PR TITLE
[MIR] Implement `as` casting (Misc cast kind)

### PR DESCRIPTION
**Pull request by nagisa**
I think that should pretty much conclude all of https://github.com/rust-lang/rust/issues/29576.

### CODE REVIEW

File: src/librustc_trans/trans/mir/rvalue.rs [hunk 1]
```diff
@@ -198,7 +198,67 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
+                                build::UIToFP(bcx, llval, ll_t_out),
+                            (Float, Int(IntTy::I)) =>
+                                build::FPToSI(bcx, llval, ll_t_out),
+                            (Float, Int(_)) =>
+                                build::FPToUI(bcx, llval, ll_t_out),
+                            _ => bcx.ccx().sess().bug(
+                                &format!("unsupported cast: {:?} to {:?}", operand.ty, cast_ty)
+                            )
+                        };
+                        OperandValue::Immediate(newval)
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/rvalue.rs]: Not sure if assuming everything is immediate here is correct, but I can’t think of a case where it wouldn’t be.

Comment (arielb1) [inline src/librustc_trans/trans/mir/rvalue.rs]: Be careful of fat pointers.

Comment (nagisa) [inline src/librustc_trans/trans/mir/rvalue.rs]: None of the cast cases written out above would handle or produce fat pointers. I’m more concerned with `Ref` vs `Immediate`. I.e. can `Misc` branch be reached with a `OperandValue::Ref` for castee.

Comment (arielb1) [inline src/librustc_trans/trans/mir/rvalue.rs]: The discriminant of an `OperandValue` is supposed to be a function of the operand's type.

**Comment by arielb1** [comment]
r+ if you add fat-pointer-to-integer and fat-pointer-to-fat-pointer casts.

**Comment by nagisa** [comment]
I added preliminary impl of misc casts from fat-ptr, but I have little idea if it is correct. E.g. currently fat-ptr → fat-ptr is just a no-op, but only because I couldn’t think of a valid fat-ptr → fat-ptr cast which would change LLVM type of the fat-ptr; I’m not sure if this is indeed correct.

**Comment by arielb1** [comment]
Ah well, I also want a test that identity casts work (e.g. `Rc<Debug> -> Rc<Debug>`). These should not be lowered to a cast at the MIR level, but that is not a reason for them not to compile.

**Comment by arielb1** [comment]
It's a separate bug.

@bors r+

**Pull request merged**